lib/system/environ/validate.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// package environ implements utilities for managing host environment
+package environ
+
+import (
+	"os"
+	"strings"
+
+	"github.com/gravitational/gravity/lib/defaults"
+	"github.com/gravitational/gravity/lib/localenv"
+	"github.com/gravitational/gravity/lib/pack"
+	"github.com/gravitational/gravity/lib/state"
+	"github.com/gravitational/gravity/lib/system"
+	"github.com/gravitational/gravity/lib/utils"
+
+	"github.com/gravitational/trace"
+)
+
+// ValidateInstall performs a local environment sanity check to make sure
+// that install on this node can proceed without issues
+func ValidateInstall(env *localenv.LocalEnvironment) func() error {
+	return func() error {
+		if err := validateNonVolatileDirectory(state.GravityInstallDir()); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := validateDirectoryEmpty(state.GravityInstallDir()); err != nil {
+			return trace.Wrap(err)
+		}
+		if err := validateNoPackageState(env.Packages, env.StateDir); err != nil {
+			return trace.Wrap(err)
+		}
+		return nil
+	}
+}
+
+func validateNonVolatileDirectory(stateDir string) error {
+	fstype, err := system.GetFilesystemForPath(stateDir)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if fstype == system.FilesystemTemporary {
+		return trace.BadParameter("installer is running from a temporary file system.\n" +
+			"It is required to run the installer from a non-volatile location.")
+	}
+	var volatileDirectories = []string{"/tmp", "/var/tmp"}
+	if os.Getenv("TMPDIR") != "" {
+		// Non-empty $TMPDIR overrides default temporary directories
+		// See https://www.freedesktop.org/software/systemd/man/tmpfiles.d.html
+		volatileDirectories = []string{os.Getenv("TMPDIR")}
+	}
+	if isRootedAt(stateDir, volatileDirectories...) {
+		return trace.BadParameter("installer is running from a temporary directory.\n" +
+			"Consider running the installer from a non-volatile location.")
+	}
+	return nil
+}
+
+func validateDirectoryEmpty(stateDir string) error {
+	empty, err := utils.IsDirectoryEmpty(stateDir)
```

Comment:
This check for empty /var/lib/gravity dir will very often fire when it's mounted on a separate disk which has default "lost+found" directory there. This will e.g. break installs with AWS provisioner. I know b/c I tried something similar in the past :) I think the check for package state that you have below is enough.

Reply:
It is not the state directory for the cluster but the operation state directory. This check verifies that the operation cannot be reinitialized once started but I realize this is poor UX - it would be much better to determine if the operation is still ongoing and do nothing - this would, for instance, involve checking whether the parameters used to create the operation are the same and if the operation is still ongoing. I will try to accommodate that.